### PR TITLE
Add 'pipeline' attribute to processor metrics

### LINKED ISSUE #10908: Processor telemetry should include Pipeline ID as attribute
**Is your feature request related to a problem? Please describe.**

The `processorhelper` package defines several metrics but provides only a "Processor ID" attribute. This is insufficiently observable because a processor ID may apply to multiple instances of a processor. In effect, all instances of the processor _configuration_ are reporting metrics as if they are the same _instance_. 

Users should be able to understand the behavior of a processor within the context of each pipeline because the usage may be very different per pipeline.

As an example, suppose a processor used in two pipelines. In one pipeline, it reports 1000/1000 log records accepted, 0 for other metrics. In the other pipeline, it reports 10/10 log records rejected, 0 for all other metrics. Currently, the metric describes this as if 99% of data is being accepted by the processor. The reality that one pipeline is very healthy while the other is very unhealthy can easily be overlooked because of the lack of a pipeline ID attribute.

**Describe the solution you'd like**

Add a "Pipeline ID" attribute to all processor helper metrics.

**Describe alternatives you've considered**

This is proposed as an alternative to https://github.com/open-telemetry/opentelemetry-collector/pull/10664, which would eliminate the need for the Pipeline ID attribute, but would introduce breaking changes elsewhere.

**Additional Context**

https://github.com/open-telemetry/opentelemetry-collector/issues/10708#issuecomment-2246340782

### PR DESCRIPTION
**Pull request by djaglowski**
Resolves https://github.com/open-telemetry/opentelemetry-collector/issues/10908

This ends up being a pretty straightforward change in actual code, but a tedious change in tests. It's also obviously a breaking change for anyone depending on these metrics.

### CODE REVIEW

**Comment by djaglowski** [comment]
One notable detail is that the `otel.signal` attribute added in https://github.com/open-telemetry/opentelemetry-collector/pull/11144 is arguably redundant, since pipeline ID's contain the data type. However, we likely will want `otel.signal` for receiver and exporter metrics as well, so perhaps it makes sense to leave it. (Connector metrics would need to differentiate between incoming and outgoing data type though.)

**Review by mx-psi — COMMENTED** [review]
The approach looks reasonable to me. There are some breaking changes on the test modules, I feel like those are okay (and that maybe we should split that part of the modules off to make our lives easier, but that's for another PR)

**Comment by djaglowski** [comment]
> The approach looks reasonable to me. There are some breaking changes on the test modules, I feel like those are okay (and that maybe we should split that part of the modules off to make our lives easier, but that's for another PR)

I added a changelog to note to breaking change in `processortest.NewNopSettings`

**Comment by djaglowski** [comment]
I opened https://github.com/open-telemetry/opentelemetry-collector/issues/11179 to capture thoughts on the broader question of which attributes to apply to each kind of component. This PR aligns with what I believe is the best overall schema, but please take a look.

File: processor/processortest/nop_processor.go [hunk 1]
```diff
@@ -20,11 +20,12 @@ import (
 var nopType = component.MustNewType("nop")
 
 // NewNopSettings returns a new nop settings for Create*Processor functions.
-func NewNopSettings() processor.Settings {
+func NewNopSettings(dt component.DataType) processor.Settings {
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply processor/processortest/nop_processor.go:23]: It seems likely we'll want to include the data type for other component kinds. I suppose we could pass the pipeline ID here for processors, but not sure there's any case to do it for other kinds. This is probably something worth trying to iron out as part of https://github.com/open-telemetry/opentelemetry-collector/issues/11179.

File: processor/internal/processor.go [hunk 2]
```diff
@@ -14,4 +17,7 @@ type Settings struct {
 
 	// BuildInfo can be used by components for informational purposes
 	BuildInfo component.BuildInfo
+
+	// PipelineID indicates which pipeline contains this processor.
+	PipelineID pipeline.ID
```
Comment (bogdandrutu) [inline processor/internal/processor.go:22]: Settings are passed to the factory method correct? Which already signals which "signal" is this for. I am confused why adding it here?

**Review by bogdandrutu — COMMENTED** [review]
Is this already done?

File: processor/internal/processor.go (hunk 2 above)
Comment (djaglowski) [reply processor/internal/processor.go:22]: This indicates the pipeline ID, not the signal. The rationale for adding pipeline ID is explained in https://github.com/open-telemetry/opentelemetry-collector/issues/10908

**Review by mx-psi — APPROVED** [review]

**Comment by bogdandrutu** [comment]
Please rebase, looks like everything changed since this change was made.

**Comment by djaglowski** [comment]
I'm closing this in favor of the graph based approach, demonstrated in https://github.com/open-telemetry/opentelemetry-collector/pull/11311#pullrequestreview-2345261928 and more robustly proposed in https://github.com/open-telemetry/opentelemetry-collector/issues/11343.

**Pull request closed without merging**